File: LEETCODE-PROBLEMS/01 - Algorithms/3 - Longest Substring Without Rep Chars - Medium/LongestSub.py
# Given a string s, find the length of the longest
# substring without repeating characters.
#
# Constraints:
#  - 0 <= s.length <= 5 * 10^4
#  - s consists of English letters, digits, symbols and spaces.


class LongestSub(object):
    def longestSub_1(self, s):
        # pointers:
        l = 0
        r = -1  # accounts for the initial window size being empty.
        
        max_length = 0
        sub = set()
        while r < len(s) - 1:
            if s[r + 1] not in sub:
                sub.add(s[r + 1])
                r += 1  # move right pointer and expand window size
                max_length = max(max_length, r - l + 1)
            else:
                sub.remove(s[l])
                l += 1
        return max_length

ls = LongestSub()
print(ls.longestSub_1('abcabcbb'))

File: LEETCODE-PROBLEMS/03 - Longest Substring Without Rep Chars - Medium/LongestSub.java
/*
 * Given a string s, find the length of the longest
 * substring without repeating characters.
 * 
 * Constraints:
 *  - 0 <= s.length <= 5 * 10^4
 *  - s consists of English letters, digits, symbols and spaces.
 */

import java.util.HashSet;
import java.util.Set;

public class LongestSub {

    /** Time complexity of the algorithm: O(n) - n is the length of the input string
     * Finds the length of the longest substring without repeating characters.     
     * @param s the input string
     * @return the length of the longest substring without repeating characters
     */
    public int longestSub(String s) {
        // pointers
        int l = 0;
        int r = -1;
    
        char[] sar = s.toCharArray();
        int max = 0;
        Set<Character> sub = new HashSet<>();
        while (r < sar.length - 1) {
            if (!sub.contains(sar[r + 1])) {
                sub.add(sar[r + 1]);
                r += 1;
                max = Math.max(max, r - l + 1);
            } else {
                sub.remove(sar[l]);
                l += 1;
            }
        } return max;
    }
}
